### PR TITLE
[Filtering] Add BasicRuleNoMatchAllRegexValidator

### PR DESCRIPTION
**Pull request by timgrein**
## Related to https://github.com/elastic/enterprise-search-team/issues/3583 ##

This PR adds the `BasicRuleNoMatchAllRegexValidator`. The default rule is excluded from this validation as it intentionally uses a match all regex.

#### Pre-Review Checklist
- [x] Covered the changes with automated tests
- [x] Tested the changes locally

### CODE REVIEW

File: connectors/filtering/basic_rule.py [hunk 1]
```diff
@@ -147,6 +147,50 @@ def from_string(cls, string):
...
+        self.validation_message = validation_message
+
+    @classmethod
+    def valid_result(cls, rule_id):
+        return BasicRuleValidationResult(
+            rule_id=rule_id, is_valid=True, validation_message="Valid rule"
+        )
+
+
+class BasicRuleValidator:
```
Comment (timgrein) [inline connectors/filtering/basic_rule.py:163]: No need to review (temporary copy from https://github.com/elastic/connectors-python/pull/268)

File: connectors/filtering/basic_rule.py [hunk 2]
```diff
@@ -147,6 +147,50 @@ def from_string(cls, string):
                 )
 
 
+class BasicRuleValidationResult:
```
Comment (timgrein) [inline connectors/filtering/basic_rule.py:150]: No need to review (temporary copy from https://github.com/elastic/connectors-python/pull/268)

File: connectors/tests/test_basic_rule.py [hunk 3]
```diff
@@ -1187,3 +1194,52 @@ def test_is_not_include_for_exclude_policy():
     )
 
     assert not basic_rule.is_include()
+
+
+def basic_rule_json(merge_with=None, delete_keys=None):
```
Comment (timgrein) [inline connectors/tests/test_basic_rule.py:1199]: No need to review (temporary copy from https://github.com/elastic/connectors-python/pull/268 until it's merged)

File: connectors/filtering/basic_rule.py [hunk 4]
```diff
@@ -147,6 +147,50 @@ def from_string(cls, string):
...
+        )
+
+
+class BasicRuleValidator:
+    @classmethod
+    def validate(cls, rule):
+        raise NotImplementedError
+
+
+class BasicRuleNoMatchAllRegexValidator(BasicRuleValidator):
```
Comment (timgrein) [inline connectors/filtering/basic_rule.py:169]: Should be reviewed

File: connectors/tests/test_basic_rule.py [hunk 5]
```diff
@@ -1187,3 +1194,52 @@ def test_is_not_include_for_exclude_policy():
...
+        (basic_rule_json(merge_with={"rule": "regex", "value": "abc"}), True),
+        # for other rule types it should be possible to use values which look like match all regexps
+        (basic_rule_json(merge_with={"rule": "contains", "value": ".*"}), True),
+        (basic_rule_json(merge_with={"rule": "contains", "value": "(.*)"}), True),
+        # invalid rules
+        (basic_rule_json(merge_with={"rule": "regex", "value": ".*"}), False),
+        (basic_rule_json(merge_with={"rule": "regex", "value": "(.*)"}), False),
+    ],
+)
+def test_basic_rule_validate_no_match_all_regex(basic_rule, should_be_valid):
```
Comment (timgrein) [inline connectors/tests/test_basic_rule.py:1241]: Should be reviewed

File: connectors/filtering/basic_rule.py [hunk 6]
```diff
@@ -147,6 +147,50 @@ def from_string(cls, string):
...
+    MATCH_ALL_REGEXPS = [".*", "(.*)"]
+
+    @classmethod
+    def validate(cls, basic_rule_json):
+        basic_rule = BasicRule.from_json(basic_rule_json)
```
Comment (wangch079) [inline connectors/filtering/basic_rule.py:174]: Should the `basic_rule_json` be validated before calling `.from_json`?

Comment (timgrein) [reply connectors/filtering/basic_rule.py:174]: No, as we'll always execute the JSON against the schema validation first. Therefore we can reuse functionality like `is_default_rule` and compare against enums without having to implement these again in this method (which would duplicate functionality we already have in `BasicRule`)

Comment (timgrein) [reply connectors/filtering/basic_rule.py:174]: No as `BasicRuleNoMatchAllRegexValidator` doesn't use a schema to validate against. The class performs a validation checking dependencies between fields dependant one the value of one field (this is not possible with fastjsonschema, you can only specify dependencies in the sense of: field X can only exist if field Y exists, without looking at the value). Therefore this validation should be separated from the schema validation.

File: connectors/filtering/basic_rule.py [hunk 7]
```diff
@@ -147,6 +147,50 @@ def from_string(cls, string):
...
+
+        if basic_rule.rule == Rule.REGEX and any(
+            match_all_regex == basic_rule.value
+            for match_all_regex in BasicRuleNoMatchAllRegexValidator.MATCH_ALL_REGEXPS
+        ):
+            return BasicRuleValidationResult(
+                rule_id=basic_rule.id_,
+                is_valid=False,
+                validation_message=f"Match all regexps: '{BasicRuleNoMatchAllRegexValidator.MATCH_ALL_REGEXPS}' are "
+                f"not allowed.",
```
Comment (tarekziade) [inline connectors/filtering/basic_rule.py:188]: I am surprised the linter did not complain (`f` not required)

**Review by tarekziade — APPROVED** [review]
LGTM

File: connectors/filtering/basic_rule.py (hunk 7 above)
Comment (timgrein) [reply connectors/filtering/basic_rule.py:188]: I think the autoformatter split it into these two lines and appended the `f`. Looks also strange to me tbh.

**Pull request merged**